app/models/user_cleaner.rb

```diff
@@ -37,17 +37,35 @@ class UserCleaner
   # Returns all users who have been inactive for INACTIVE_USER_THRESHOLD months,
   # i.e. their last sign-in date is more than INACTIVE_USER_THRESHOLD months ago.
   #
-  # Users without a last_sign_in_at date are also considered inactive. This is
+  # Users without a current_sign_in_at date are also considered inactive. This is
   # the case for users who have never logged in since PR #553 was merged.
+  #
+  # Edge cases for registration (that refine the above statements):
+  # - A user might have registered but never actually logged in (confirmed their
+  #   email address). In this case, we don't look at the current_sign_in_at date
+  #   (as this one is still nil), but at the confirmation_sent_at date to
+  #   determine if the user is considered inactive.
+  # - Another edge case is when users have registered and confirmed their mail,
+  #   but never logged in after that. In this case, current_sign_in_at is indeed nil,
+  #   but the user should only be considered inactive if the confirmation_sent_at
+  #   date is older than the threshold.
   def inactive_users
-    User.where(last_sign_in_at: ...INACTIVE_USER_THRESHOLD.ago)
-        .or(User.where(last_sign_in_at: nil))
+    User.where.not(confirmed_at: nil).and(
+      User.where(current_sign_in_at: ...INACTIVE_USER_THRESHOLD.ago)
+      .or(User.where(current_sign_in_at: nil,
+                     confirmation_sent_at: ...INACTIVE_USER_THRESHOLD.ago))
+    ).or(
+      User.where(
+        confirmed_at: nil,
+        confirmation_sent_at: ...INACTIVE_USER_THRESHOLD.ago
+      )
+    )
   end
```

Comment:
I totally agree that readability and maintaianability is more important that performance here.
If you wanted to make it even more readableyou could do the following:
Add the following (or similar) scopes on the `User` model:

```ruby
scope :confirmed, -> { where.not(confirmed_at: nil) }
scope :unconfirmed, -> { where(confirmed_at: nil) }
scope :no_sign_in_data, -> { where(current_sign_in_at: nil) }
scope :inactive_since, ->(threshold) { where(current_sign_in_at: ...threshold.ago) }
scope :confirmation_sent_before, ->(threshold) { where(confirmation_sent_at: ...threshold.ago) }
```
But if you want to keep it as it is, feel free to resolve without code change.
Then this rewrites as
```ruby
def inactive_users
  threshold = INACTIVE_USER_THRESHOLD
  User.confirmed.and(User.inactive_since(threshold).or(User.no_sign_in_data.confirmation_sent_before(threshold))
  ).or(User.unconfirmed.confirmation_sent_before(threshold))
end
```

Reply:
Thanks, that reads so much better!

The only thing: I don't want to put this into the `User` model, it's bloated enough. Right now, we only use your proposed scopes inside the user cleaner, so why not incorporate them there in some way. E.g. I could imagine a `user_inactivity.rb` class that does `@user = user` in its `initialize` method and then implements the scopes as methods like `inactive_since(...)` etc.

I'm just not sure where to place this file 😅 In a clean architecture, I'd place it in the _application layer_ and the job that is scheduling the user cleaner in the _infrastructure_ layer. While the mapping `job/ -> "infrastructure layer"` is quite clear, I'm missing the "application layer folder" in the MaMpf codebase. Maybe I should just put the User Cleaner and the `user_inactivity` class inside the `worker` for now?

Reply:
I do not know how the idea with scopes as methods in user_inactivity.rb will work. Scopes are Active Record queries (in the above code, for the `User` model). They are not called on an instance of `User`, but rather on the `User` class. Also, they can be chained as above. If you were to implement them as methods in user_inactivity.rb, I do not see how they become such queries. But of course that does not mean that it is not possible ;-).

Reply:
After a night's sleep, I think maybe this is just the wrong PR to discuss about bloated models and architecture etc. The discussions we will have for #694 will lead the way if and how to refactor the codebase and possible architectural changes. But as of now, we have not yet made a decision. Yes, the User model is bloated. But it is bloated because of reasons that have nothing to do with #693, not because of a possible adding of scopes. In vanilla Rails Logic, the above scopes belong to the `User` model (while other stuff should probably be extracted form there, I guess). Just as UserCleaner is a model as of now and belongs to the model folder (in #671 I introduced the `app/services` folder, maybe it could be placed there in the future if the changes in #671 should go through).

Reply:
Yeah, you're right that is this probably the wrong place to discuss this ;) I added the changes in 70c65fc19e